### PR TITLE
feat: implement new interface

### PR DESCRIPTION
**Pull request by rsmayda**
### `verifyAccessToken` changes
- Expected `fhirUser` regex is now a constant `FHIR_USER_REGEX`
  - `fhirUser` is expected to be an absolute url
  - If `fhirUser` claim is not in the right format we reject
- metadata & .well_known changed to pass authorization
- return the userIdentity found by the call to userInfo

### `authorizeAndFilterReadResponse` changes
- For searching filter out resources user does not have access to, does NOT throw error
  - Thought here is if all elements are filtered out the router should call for the next page of search results and try again
- For single reads, throw error if user is not authorized
- For external users (ie their resource is not from our FHIR server) they must use the fully qualified reference to have access
- To check if user is a reference I am converting the JSON to a string then checking for the `fhirUser` claim
- Internal Practitioners can access anything

### `isWriteRequestAuthorized` changes
- Make this Practitioner only action

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/smartHandler.ts [hunk 1]
```diff
@@ -5,36 +5,61 @@
...
-import { LaunchType, ScopeType, SMARTConfig } from './smartConfig';
+import { IdentityType, LaunchType, ScopeType, SMARTConfig } from './smartConfig';
 
 // eslint-disable-next-line import/prefer-default-export
 export class SMARTHandler implements Authorization {
     static readonly CLINICAL_SCOPE_REGEX = /^(?<scopeType>patient|user|system)\/(?<scopeResourceType>[a-zA-Z]+|\*)\.(?<accessType>read|write|\*)$/;
 
     static readonly LAUNCH_SCOPE_REGEX = /^(?<scopeType>launch)(\/(?<launchType>patient|encounter))?$/;
 
+    static readonly FHIR_USER_REGEX = /(?<hostname>(http|https):\/\/([A-Za-z0-9\-\\.:%$]*\/)+)(?<resourceType>Person|Practitioner|RelatedPerson|Patient)\/(?<id>[A-Za-z0-9\-.]+)$/;
```
Comment (nguyen102) [inline src/smartHandler.ts]: Can we add some tests for this REGEX? It'll help us document what we expect this REGEX to match. It'll also confirm that the REGEX works as expected.

File: src/smartHandler.ts [hunk 2]
```diff
@@ -54,53 +79,118 @@ export class SMARTHandler implements Authorization {
...
+
+    private getFhirUser(userIdentity: KeyValueMap): { hostname: string; resourceType: IdentityType; id: string } {
+        const { fhirUserClaimKey } = this.config;
+        const fhirUserValue = userIdentity[fhirUserClaimKey];
+        const match = fhirUserValue.match(SMARTHandler.FHIR_USER_REGEX);
+        if (match) {
+            const { hostname, resourceType, id } = match.groups!;
+            return { hostname, resourceType, id };
+        }
+        throw new UnauthorizedError("Requester's identity is in the incorrect format");
```
Comment (nguyen102) [inline src/smartHandler.ts:192]: Do we want to pass along the `userIdentity` in the error message for context?

File: src/smartHandler.ts [hunk 3]
```diff
@@ -54,53 +79,118 @@ export class SMARTHandler implements Authorization {
...
+        return R4_PATIENT_COMPARTMENT_RESOURCES;
     }
 
     async authorizeAndFilterReadResponse(request: ReadResponseAuthorizedRequest): Promise<any> {
-        // TODO this is stubbed for now
-        await this.isAuthorized({ accessToken: request.accessToken, operation: request.operation });
-        return request.readResponse;
+        const fhirUser = this.getFhirUser(request.userIdentity);
+        const { operation, readResponse } = request;
+        if (this.searchOperations.includes(operation)) {
```
Comment (nguyen102) [inline src/smartHandler.ts:132]: What is the list of `searchOperations` for, are those operations that we support for SMART? Should that a be a config value, or is it always a set list?

Also the list does not include the `read` operation. Does that mean a requested won't be able to read any resources?

[Reference in Router](https://github.com/awslabs/fhir-works-on-aws-routing/blob/mainline/src/router/routes/genericResourceRoute.ts#L40)

**Review by nguyen102 — DISMISSED** [review]
Looks good. Some minor comments.

File: src/smartHandler.ts (hunk 1 above)
Comment (rsmayda) [reply src/smartHandler.ts]: +1 good call out and practice; found some minor additions

File: src/smartHandler.ts (hunk 2 above)
Comment (rsmayda) [reply src/smartHandler.ts:192]: I am hesitant as it may contain PII -- I think better to log/return less

**Review by nguyen102 — APPROVED** [review]

**Pull request merged**